### ISSUE TITLE
Dear ImGui licensing

### ISSUE THREAD

**Issue by Aspie96** [ocornut/imgui#6076]
(I am not a lawyer and this is not legal advice).

The license of Dear ImGui is the MIT license, which requires licensees to keep the copyright notice by licensor.
In general, when a contribution is given to an open source project the way it usually happens on GitHub, the contributor owns and is the licensor of that contribution, while the maintainer (and everyone else) is the licensee, so everyone has to comply with the text of the license.

In commit 05ec0b07365b36319ed6ef301c723634c7f8777d, created after issue #1346, the mention of "ImGui contributors" was removed from the copyright notice.
In 814ecedd1e86400a4ddb0ac468c8ceed9acc1103, created after issue #5337, a CLA was added to the project, requiring contributors to transfer their copyright on contributions to the maintainer.

If copyrightable contributions to Dear ImGUI from other parties exist which predate 814ecedd1e86400a4ddb0ac468c8ceed9acc1103, wouldn't they still be owned by the contributor?

My concern is specifically about contributions prior to 05ec0b07365b36319ed6ef301c723634c7f8777d, since the copyright notice, which should have been preserved, has been altered since then.

Does the codebase contain any significant portion of copyrightable contributions received prior to 05ec0b07365b36319ed6ef301c723634c7f8777d which are still owned by parties other than @ocornut?

Am I missing something? Thank you a lot in advance!

**Comment by JustaSimpleUser** [comment]
And, where is the license posted?

**Comment by Aspie96** [comment]
It's the LICENSE.txt file in the root of the repository.

**Comment by JustaSimpleUser** [comment]
Thanks

**Comment by ocornut** [comment]
> Does the codebase contain any significant portion of copyrightable contributions received prior to https://github.com/ocornut/imgui/commit/05ec0b07365b36319ed6ef301c723634c7f8777d which are still owned by parties other than @ocornut?

I don’t know, it’s probably difficult to evaluate. There is a common grey area with I think a majority of permissively licensed software which didn’t have a CLA in place. It’s equally rare to see eg a MIT licence where copyright is attributed to dozens of people. And it is very unlikely that people would effectively care given there’s no direct financial gain at stake from owning the copyright itself. Even more so that in the event or someone would claim a significant commit prior to that date it would have been altered a lot since, and we would simply rewrite it if required.

So I don’t know what to answer you. You are right but its a very common grey area and its fixable in the sense that isn’t much significant code that hasn’t been vastly rewritten and iterated dozen of times. I believe there’s nothing to do.

**Comment by Aspie96** [comment]
> There is a common grey area with I think a majority of permissively licensed software which didn’t have a CLA in place.

Usually there is not much of a gray area.
Each individual author holds the copyright to their own contributions and everybody licenses all the code to everybody else under the same license.

> It’s equally rare to see eg a MIT licence where copyright is attributed to dozens of people.

Copyright is automatic and comes with authorship. It does not require a copyright notice to exist.
The default is that the author retains copyright (whether it's recognized in a copyright noticed or not). But authors can transfer it, for example trough a CLA.
That said, it's actually not rare to see "authors" or "contributors" in a copyright notice (even where none is mentioned individually).

Contributions between 814ecedd1e86400a4ddb0ac468c8ceed9acc1103 and 05ec0b07365b36319ed6ef301c723634c7f8777d are also still copyrighted by the authors (since the CLA didn't exist yet), but this is not a problem, since the license at that point didn't require any copyright notice beyond that referring to you.

The concern is whether a user of the library needs to re-add the remove copyright notice in order to comply with all conditions from all licensors. I don't think it's just about possible legal risk either, but rather about not infringing upon the rights of authors.

I ran the following command:
`git ls-files | awk '{system("git blame " $0 " -ce --abbrev=40")}' | awk -v FS='\t' -v OFS='\t' '$2 !~ /(omarcornut|ocornut)@/ && $2 {print $1, $2}' | awk '!seen[$1]++' | awk 'system("git merge-base --is-ancestor " $1 " 05ec0b07365b36319ed6ef301c723634c7f8777d") == 0'`

Which showed the presence of code still in the repository (never rewritten or replaced) which was committed prior to 05ec0b07365b36319ed6ef301c723634c7f8777d by contributors other than yourself.

I think it can be useful to ask for their feedback. I excluded those who had only made one one-line commit prior to 05ec0b07365b36319ed6ef301c723634c7f8777d):
- @JuJuBoSc (author of 1f3a0e78ccab65519ea4122c24c40715f057554b)
- @nlguillemot (author of d4d51a78027fa9e44da46fc712d05216a8d0186a)
- @josiahmanson (author of 5fe2cacd4da495be9b656b29d6e2e848434f447a)
- Trezanik (author of 790275eae234bcaef1c94dcc6d54ab6090671f68)
- @Loftilus (author of 4cbd316f0183600222625a9da46ab4786bfe2981, 08ef9819c8b6e7fa21f5acccc57a4e6692b1dcc0, 1394616d9c87114d5c84b1deb6d75a58dc82d504, 99b4f951b91a35fe7ad9a3b0e81961da3fc25e00 and 0ceddc29ff5b80dd270b7810f6bfa931c8353e7c)
- @martty author of 53126cd7137bdc0bbedbe47ec7df2ee2d4ef05f5, 509df3e2790f9fb6162d9cb2a4f2f3dd6465fdb8 and 4bc6a951fe11d34c0e64af08d1802aef011f2215)
- @FL1NE (author of 9bb95db68a8bfe2acdd862113dbb38d39d6f8f09)
- @fr500 (author of 8b6896faf9bf590cb2e88e9e914b10d1e46b4ef8)
- @gordonmcshane (author of 6ee65fc5afb6a0f14c88ec3d693d5fc8a2fcfb52)
- @ParticlePeter (author of 56dff3a0803de0450ea242dd0757d2f51d9eb75a, 33874073dc3275a7d99e20c9986d39998328b62f and 201d589714c6ef7ff40c87140818a2208b8686b7)
- @CedricGuillemet (author of 2acdafe4f2cf844c6e21d77e83b605b7aa1aa75c)
- Zach Toogood (author of e130b09d9182641d7ca4995b5cfd6bc0d5ca9852)
- @joeld42 (author of eee6dab226b2f056598d621aeed16691bbce7fe7)
- @mmozeiko (author of 2142181c515b8a985bec075ee589c7b76298b9fa)
- @thedmd (author of d75d2b1871831377faef836f4be798575732e412, 7588dfb67e7dd8970b64bffc87c2659fe7684ea5, 79f07f6ff040f6bfbf6b86320bf417f1d495966b and 8b2d449b165f2214a088f8d0cd8a20b684e99787)
- @bkaradzic (author of a59a04f4d0822b69150433fb7cbf70e944874c07)
- @sergof (author of 6e8579fc144f1d59a75aabf3740470d3e3739a1b and d5eb87d0a22b1232ef57d5ecf742cd3d9c53c437)
- @DanielGibson (author of 02a09672434295b657274edba8969a1d4ca3f18d and be8fb858cc810ff2125669e4e7e1254405f1ebc8)
- @septag (author of 25080d53e5bed0203fe5182e78907668e4080a77)
- @emoon (author of c9c41c38747921162ece62e4e151dc71ba13c9f3)
- @elect86 (author of 9e3f9fc2ddf0a170788f6b101c5027f82dd35cde, db190c16f6a970fdaa918ae6be97bb4b246b20b7 and 1b2ec35b8ded8399cc3d47642121e090ef353f65)
- @kylawl (author of 3f46d9b9337e7ed0f5526f7f4c4dfda2e5c2820f and b8fcb4e7e450d536c2e881ae026a8f3227219865)
- Dale Kim (author of 85c71170affc91c4cc396623b54ef252cf2c49ee)
- @tpoechtrager (author of 3604aecf8fa6465f19475fab117336b41be5261c)
- @pdoane-blizzard (author of 70cb427469de3da6e0d0db8b806c33b2899f08bf)
- @zapolnov (author of a908c109c09cbb54962bb5b3a5056c188fe260e7)
- @cosmy1 (author of 4b5a4cae09d0efd32c64b510490672d35bbccb51)
- @ecraven (author of 659b3fd259c62820ec5c133da48214052d23213b)
- @gtucker (author of 66c428129051b278d252de9b1ec66443e49d48da)
- @osman-brian (author of 69e6f299f32166a7557861b95e452bab8a16ecdf)
- @inolen (author of c38526d14b38eff269032ac2b00a748ee16964c6)
- @zho7611 (author of cc53d317646a32289fab7d6539809a0bb6329dd4)
- @nProtect (author of 4141144b217ce89e819a1c9a4aac165789342188)

While this list is non-exhaustive (because the method I used will not reveal all affected code), it would be interesting to know from these authors if they wish to transfer copyright to you, and/or if they are ok with only the new copyright notice being used.

Given the information available to me so far, it seems reasonable for any user of the library to re-add the removed copyright notice.

**Comment by inolen** [comment]
I do not care re: my tiny commit.

With that said, I don't really understand the purpose of this issue. You can clearly see ocornut is the primary author of this code, what's the purpose of laying out all of these legal concerns and pinging so many people when you're in fact not a lawyer and are not attempting to provide legal advice.

This feels like yet another example of the GitHub-ification of open source, where in project authors are exhausted by sideways issues from people who are not contributing value to the project.

**Comment by mmozeiko** [comment]
I have done only minor contribution to imgui codebase, but I don't care about exact licensing terms. Omar can change it to whatever he wants. When I decided to contribute to this codebase I did that because I found code useful and interesting. I did not that do because I liked (or disliked) licensing terms.

Btw, I find this kind of ping in issue unrelated to commit from non-author of project quite rude.

**Comment by fr500** [comment]
My contribution is a build fix, I couldn't care less for that.

I love the fact that it's so permissive, but at the same time I think Omar should have the right to relicense the code in any way he thinks, he has built this amazing library for free in his spare time. Many many projects use it, and no matter what you can always go back to the MIT implementation of a very functional library.

The CLA only affects code that is pulled back into the main repository and that comes from an intent to get a fewture merged (pull request). You can change what you want and keep it to yourself under your own copyright if you desire so. So this is completely irrelevant.

**Comment by zapolnov** [comment]
My changes were super tiny, do whatever you want with them 😄

**Comment by josiahmanson** [comment]
Same here. Do whatever you want with my changes.

**Comment by PathogenDavid** [comment]
> Usually there is not much of a gray area.

This is an absurdly common gray area affecting tons of open source projects, especially older ones like Dear ImGui. If you don't think it is you haven't been paying attention.

Also it's extremely rude to `@` so many people, especially when you aren't involved in this community at all.

If a contributor to Dear ImGui takes issue with the copyright notice, let them bring it up themselves.

You really think you know more about the legal issues involved here than the legal council at all the companies already using Dear ImGui? You're just some dude, and a rude one at that.

**Comment by DanielGibson** [comment]
I don't care about the license for my contribution, because it was tiny.

If I had cared about my name being credited I'd have added it to the license header of the files I touched.

Also, IMO it doesn't make a real difference if the line in the license header says "(C) Omar Cornut" or "(C) Omar Cornut and ImGui contributors", because either way my name isn't mentioned anywhere (but the git commit history)..
*(Would be slightly different if there had been a global Contributors.txt or something that lists everyone who ever contributed, which would credit the contributors without bloating the license header of the source files)*

**Comment by Aspie96** [comment]
Given the overwhelmingly negative response to this issue, and also given to the realization that "fixing" it without adding back the original copyright notice (something which doesn't seem to be wanted) is likely close to impossible given the sheer amount of authors, I've decided to close this as "wontfix", as it sounds accurate, and to amend my comment by removing the authors list.

I will simply add back the original copyright notice when I use the library and make sure to report this every time I advise others to use the library.

I apologize for any annoyance caused. It was not my intention. When I opened the issue I didn't know if I had missed something (such as a rewrite of code or some collective transfer of copyright, both of which do sometimes happen). After Omar's response I should probably have investigated this privately, and then reached the conclusion which I have reached anyways.

I will still respond to this, however.

> This is an absurdly common gray area affecting tons of open source projects, especially older ones like Dear ImGui. If you don't think it is you haven't been paying attention.

The fact that authors own copyright by default is a direct consequence of copyright law. The assumption that contributions are licensed (and without a license they couldn't be used) under an "inbound=outbound" licensing model is the [typical](https://opensource.com/law/11/7/trouble-harmony-part-1#contribution-policies-and-free-software-tradition) assumption in open source projects. This is also the interpretation of the [Google  Open Source Casebook](https://google.github.io/opencasebook/authorship/#section-3---writings-the-role-of-contributor-license-agreements-developer-certificates-of-origin-and-other-writings-in-governing-copyright-ownership-of-contributions-to-open-source-projects). On GitHub in particular, this is clearly written down in the terms of service and [has been](https://web.archive.org/web/20171202034216/https://help.github.com/articles/github-terms-of-service/) since earlier than 05ec0b07365b36319ed6ef301c723634c7f8777d.

**Comment by GamingMinds-DanielC** [comment]
Usually I don't care about what other projects people are involved in, unless they post things that are particularly interesting to me (good way to stumble on useful projects). Or if they post suspicious stuff on the other end. On your activity page it states that you contributed to a repository called "github-dramas". I don't know if that was your actual intention, but it certainly makes it look like you tried to create tension or drama for the sake of it. Not a good look.

**Comment by ocornut** [comment]
To clarify my thoughts on this: in spite of the tension here you are raising a reasonable point and I appreciate the intent.
I just don't think there is a meaningful action to perform.
Ironically and amusingly, mass-tagging people moved the needle forward a little bit: it exhibit a common sentiment, and it gave a chance for 6 persons to explicitly give their OK. Were we to find/chase the other 50+ I am confident an overwhelming majority wouldn't mind transferring copyright, and in the case it was needed for the project (but I don't think it will) and some people didn't answer or didn't answer positively, it would be also reasonable on my end to decide to rewrite them. In fact, for the sake of it, I am now going to ask the 10 top contributors about this.

**Comment by Clownacy** [comment]
The simple solution to this is to just re-add the 'and ImGui contributors' to the license. The MIT licence says it in plain English:

```
The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.
```

It costs nothing, and it keeps past contributors from having a reason to DMCA the repo.

**Comment by thedmd** [comment]
I'm fine with this. I contributed my work to help with the project and I consider such code to belong to the project for anyone to use according to project license.

**Comment by ocornut** [comment]
I should also mention that we can and we should add a CONTRIBUTORS.txt file, essentially reflecting the list available at https://github.com/ocornut/imgui/graphs/contributors

The short form that has been in the README for a long while links to that list:
"Developed by [Omar Cornut](https://www.miracleworld.net/) and every direct or indirect [contributors](https://github.com/ocornut/imgui/graphs/contributors) to the GitHub."

It would be nice if the CONTRIBUTORS file gives a little more details for non-trivial contributions, will see how I can get that sorted out.

**Comment by Aspie96** [comment]
First, I want to thank @ocornut for his balanced response (and of course for Dear ImGui).

> Were we to find/chase the other 50+ I am confident an overwhelming majority wouldn't mind transferring copyright,

This was of course my prediction as well.

Indeed, they wouldn't even need to transfer copyright.
Everything between 05ec0b07365b36319ed6ef301c723634c7f8777d and 814ecedd1e86400a4ddb0ac468c8ceed9acc1103 is copyrighted by contributors (since there was no CLA), but without the old copyright notice.
Even if there were contributors unwilling to transfer copyright, I feel almost certain all of them would be ok with the new copyright notice.

The reason for the issue was absolutely not a prediction to the contrary. Rather, I think how I use code should be guided on the actual license authors used, not my prediction of what they would be ok with.

While I think the issue did deserve to be noted (if anything, for those who, like myself, wish to comply with license terms strictly the way @Clownacy described), I still regret "moved the needle forward a little bit", since in retrospect I clearly see reaching out to all past contributors is effectively impossible, and hence I renew my apology.

> I should also mention that we can and we should add a CONTRIBUTORS.txt file, essentially reflecting the list available at https://github.com/ocornut/imgui/graphs/contributors

While many project have this, and maybe contributors would like it, this is of course not necessary for the issue at hand.

**Comment by podsvirov** [comment]
I make my changes in order to improve the project for myself and I hope that it is useful for others. My changes became part of the project.
I agree to transfer copyright of [my commits](https://github.com/ocornut/imgui/commits?author=podsvirov) to @ocornut.

**Comment by ocornut** [comment]
Quick update on this:
I've contacted or checked with 38 peoples who are "top-contributors" (mixture of line count and commit count).
For reference, under a threshold of 40-45 people (still working on contacting more, but some contacts are difficult to find) the contribution are under ~55 lines.

So far out of the 38 contacts I have received 28 answers, 27 positive and 1 negative. 
Thanks for answering!

Bartosz also sent me:
> Recently I have been reading about what is and what isn't copyrightable, for example this EFF primer about the Open Game License: https://www.eff.org/deeplinks/2023/01/beware-gifts-dragons-how-dds-open-gaming-license-may-have-become-trap-creators
> I do not believe my changes in ImGui are copyrightable. However, if you feel they are, be free to assume I have transferred the copyright to you.

_"Courts are essentially interested in whether the uncopyrightable elements of the work remain available for the public to use, or if the copyright owner is effectively monopolizing them because there simply aren’t many different ways to describe the uncopyrightable ideas or system."_

This is interesting because it may cover the case that contributions in particular minor ones are not copyrightable in the first place. Hypothetically e.g. simple additions, typos corrections would fail under this but I'll need to do more research and I'll keep contacting people.

**Comment by Aspie96** [comment]
> So far out of the 38 contacts I have received 28 answers, 27 positive and 1 negative.

I was not optimistic about so many people responding, this is going quite well, and better than I anticipated.
I am surprised by the one negative response.

Out of curiosity, are you basing yourself on contributions prior to 05ec0b07365b36319ed6ef301c723634c7f8777d (just for the copyright notice) or to 814ecedd1e86400a4ddb0ac468c8ceed9acc1103 (for full transfer of copyright)?

> This is interesting because it may cover the case that contributions in particular minor ones are not copyrightable in the first place.

It's likely that some contributions are copyrightable and some aren't, but also that this varies widely by jurisdiction. For example, different jurisdictions have different [thresholds of originality](https://commons.wikimedia.org/wiki/Commons:Threshold_of_originality).

Nonetheless, I fully agree that contributions which are cumulatively (over all one's contributions) trivial or quite small or simple can generally be safely assumed not to be original enough to be copyrightable.

I think Bartosz's quote of the EFF was, however, specifically referring to the [merger doctrine](https://en.wikipedia.org/wiki/Idea%E2%80%93expression_distinction#Merger_doctrine) of the US.

**Comment by ocornut** [comment]
> Out of curiosity, are you basing yourself on contributions prior to https://github.com/ocornut/imgui/commit/05ec0b07365b36319ed6ef301c723634c7f8777d (just for the copyright notice) or to https://github.com/ocornut/imgui/commit/814ecedd1e86400a4ddb0ac468c8ceed9acc1103 (for full transfer of copyright)?

I sent everyone the list of all their commits regardless of date.
Thank you for the link about Threshold of Originality.
Bartosz sent me the EFF link but that specific quote is one I highlighted. May have overlooked other interesting element of that article.

I'll keep this updated when I can but for now going back to coding, as I already probably spent many hours on this this week.

**Comment by samhocevar** [comment]
I have read this discussion but have not found enough information on why a CLA was necessary or why it is a good thing. In general I dislike them.
 - they mean I cannot relicense my own code or contribute it to another project under a different license
 - if the license is violated by an entity and the sole copyright holder is unwilling to take action (whether by laziness, lack of interest, but maybe also because they are being bribed), the other contributors have no course of action

For years the tech and opensource community have complained about Red Hat’s CLAs and they eventually stopped using them. The GCC steering committee stopped requiring copyright transfer last year. My contributions to Dear ImGui are not very significant, and I’m not going to try and block the whole process, but you should still know that it creates a problem for me and the rest of the world while it doesn’t solve any that I can see.

**Comment by fr500** [comment]
Why would it create a problem with a permissive license though?

You only transfer copyright when you want your code pulled into mainline, if someone is building some customization for themselves, they can keep it in their fork and retain full copyright. 

Also if anyone, not just ocornut wants to use Dear IMGUI in a closed source project... they can already. It can be used anywhere without pretty much any restriction.

I mean, yeah the CLA can be annoying, but in this particular case... it achieves absolutely nothing unless the library were to be relicensed, and there is literally no reason to relicense since the license is so open already.

**Comment by FunMiles** [comment]
I have no problem transferring my contribution copyright to @ocornut . I am, however interested in what @samhocevar mentions. I also maintain a small project for raspberry pi pico (https://github.com/FunMiles/PicoAsync) and would face similar issues. I have not picked a license for it yet and so far, I am the only contributor, but if people start contributing, I'll need to pick a license and understand it.

**Comment by Aspie96** [comment]
@FunMiles, I do not know exactly what you are missing, but I will do my best to simplify and summarize, as I think it might also be helpful to a reader of this thread.

Usually copyright (which is a set of exclusive rights) comes with authorship and any other party will need a license to use ("use" including distributing and modifying) the work. Authors use open source licenses to allow the public to use their work in ways which would otherwise be reserved to them. Open source licenses (with a couple exceptions) have conditions which the user needs to follow for the license to be valid.

The common convention in the open source community is that contributors hold the copyright to their own individual contributions. In the specific case of GitHub projects, this default is also expressly clarified in the terms of service. All contributors use the same open source license which represents a direct permission from each one of them to the user, all under the same conditions.

As an alternative, the maintainer of a project may ask contributors to sign an agreement trough which the contributor transfers their rights on the contribution to the maintainer. In this case, one only needs a license from the maintainer, not anyone else, to use the project, since there is no other copyright holder.

In the symmetric "inbound=outbound" scenario, the terms of the license used for past contributions need to be followed forever by all users (including other contributors, and even the maintainer), unless relicensed by all past contributors. In the asymmetric case where the maintainer holds all copyright, the maintainer can trivially relicense the project under a new license.

A project in the "symmetric" scenario can very much move to a new open source license for future contributions, if the two are compatible, but will still have to comply with the terms of the past license as well.

From the point of view of the end user, there is basically no difference between the two scenarios, but Contributor License Agreements which transfer copyright to the maintainer are controversial for many reasons.

What license to pick for your own projects is really up to you and depends on your own goals. The MIT license (the one used by Dear ImGui) is rather simple: the user only needs to preserve a verbatim copy of the text of the license itself and the copyright notice(s) above it in every copy of the program (modified or not, full or partial).
